packages: Add dev keywords for Composer

anomiex opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
Modern versions of Composer will automatically suggest passing `--dev` to `composer require` if certain keywords ("dev", "testing", or "static analysis") are set on the package. Let's set those keywords.

And then we may as well set additional keywords on them too since we're doing it.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Proofread, see if the keywords chosen make sense.
* Not sure the `--dev` recommending can be tested without this being merged and released. 🤷

anomiex committed:
packages: Add dev keywords for Composer

Modern versions of Composer will automatically suggest passing `--dev`
to `composer require` if certain keywords ("dev", "testing", or "static
analysis") are set on the package. Let's set those keywords.

And then we may as well set additional keywords on them too since we're
doing it.

anomiex committed:
Change entries and version bumps

anomiex committed:
Update lock files again

anomiex commented:
Although... within the monorepo we still need to have jetpack-autoloader and jetpack-composer-plugin as non-dev so the build can continue to use `--no-dev` and still get the autoloader and jetpack_vendor stuff. Sigh.

jeherve reviewed (commented):
I think it makes sense for changelogger and codesniffer, but I don't know if I would suggest `dev` for the autoloader or `composer-plugin`, since those need to be included in production builds for consumer plugins as far as I know.

anomiex commented:
> but I don't know if I would suggest `dev` for the autoloader or `composer-plugin`, since those need to be included in production builds for consumer plugins as far as I know.

It's a little complicated. We don't actually need the autoloader package itself (`vendor/automattic/jetpack-autoloader/`) in production builds, just the results of it (`vendor/jetpack-autoloader/` and `vendor/composer/jetpack_autoload_*`). Same goes for composer-plugin.

But to get those effects the package has to be installed, so it won't work to have them in `require-dev` while doing `--no-dev` to try to prepare for bundling composer deps. So yeah, it might be better to remove the "dev" keyword for those.

anomiex committed:
Merge remote-tracking branch 'origin/trunk' into add/composer-dev-keywords

anomiex committed:
Remove dev keywords from autoloader and composer-plugin

We need them installed even with `--no-dev` so their side effects still
happen. Probably everyone else will need that too.
